.github/workflows/rust.yml

```diff
@@ -19,26 +19,41 @@ jobs:
           default: true
           override: true
           components: clippy, rustfmt
+      - uses: Swatinem/rust-cache@v1
+        with:
+          key: ${{ runner.os }}-checkout
 
       - name: Cancel previous runs
         uses: styfle/cancel-workflow-action@0.5.0
         with:
           access_token: ${{ github.token }}
 
+      - name: Build
+        run: cargo build --all --release
+      - uses: Swatinem/rust-cache@v1
+        with:
+          key: ${{ runner.os }}-build
+
       - name: Run tests
         run: cargo test
+      - uses: Swatinem/rust-cache@v1
+        with:
+          key: ${{ runner.os }}-test
 
       - name: Check formatting
         run: cargo fmt  -- --check
+      - uses: Swatinem/rust-cache@v1
```

Comment:
May not matter with as fast as I’ve seen rust tests run so far but if the test suite ever gets longer we probably wanna run fmt and possibly clippy first since those are more or less constant time tasks and failing those first will return the result to be fixed faster than if we wait on any long test suites to pass only to fail on a format or linter error

Reply:
> we probably wanna run fmt and possibly clippy first since those are more or less constant time tasks and failing those first will return the result to be fixed faster

Well according to [this](https://www.reillywood.com/blog/rust-faster-ci/) blog, clippy can reuse build artifacts (which is why I flipped the order)

Reply:
Sorry, I meant “first” after the build step. So:
- build
- fmt
- clippy
- test
- If/then package

Reply:
Sure, updated in bf11069